File: aicp-frontend/src/views/canvas/director/state/directorDocument.js
/**
 * 导演台文档协议（前端侧）。
 * 坐标：RH_Y_UP_METERS。旋转：归一化 Quaternion。时间：半开 [0, duration_ms)。
 */

export const COORDINATE_SYSTEM = 'RH_Y_UP_METERS'

/** 帧索引计算：frame_count = ceil(duration_ms / 1000 * fps) */
export function frameCount(durationMs, fps) {
  return Math.ceil(durationMs / 1000 * fps)
}

/** 有效帧范围：{ first: 0, last: frameCount - 1 } */
export function validFrameRange(durationMs, fps) {
  const count = frameCount(durationMs, fps)
  return { first: 0, last: count - 1 }
}

/** 时间(ms) → 帧索引 */
export function timeToFrame(timeMs, fps) {
  return Math.floor(timeMs / 1000 * fps)
}

/** 帧索引 → 时间(ms) */
export function frameToTime(frame, fps) {
  return Math.round(frame * 1000 / fps)
}

/** 创建空的 DirectorDocument */
export function createEmptyDocument(durationMs = 5000, fps = 24, aspectRatio = '16:9') {
  return {
    coordinateSystem: COORDINATE_SYSTEM,
    durationMs,
    fps,
    cameras: [{
      id: 'camera_1',
      name: '主相机',
      focalLengthMm: 50,
      sensorWidthMm: 36,
      aperture: 2.8,
      nearClip: 0.1,
      farClip: 1000,
      aspectRatioOverride: aspectRatio
    }],
    activeCameraId: 'camera_1',
    objects: [],
    timeline: { tracks: [] }
  }
}

/** 检查是否应自动保存 */
export function shouldAutosave(state) {
  return state.dirty && !state.validating && !state.frozen
}

/** 从文档获取活跃相机 */
export function activeCamera(doc) {
  return doc.cameras?.find(c => c.id === doc.activeCameraId) || doc.cameras?.[0] || null
}

File: aicp-backend/src/main/java/com/aicp/module/director/domain/DirectorDocument.java
package com.aicp.module.director.domain;

import com.fasterxml.jackson.annotation.JsonInclude;
import java.util.List;

/**
 * 导演台文档协议 v1。
 * 坐标系统：RH_Y_UP_METERS。旋转：归一化 Quaternion。
 * 时间区间：半开 [0, duration_ms)。
 */
@JsonInclude(JsonInclude.Include.NON_NULL)
public record DirectorDocument(
        String coordinateSystem,   // "RH_Y_UP_METERS"
        int durationMs,
        int fps,
        List<CameraDefinition> cameras,     // 1–8 个相机
        String activeCameraId,
        List<SceneObject> objects,
        Timeline timeline
) {
    public DirectorDocument {
        if (coordinateSystem == null) coordinateSystem = "RH_Y_UP_METERS";
        if (cameras == null) cameras = List.of();
        if (objects == null) objects = List.of();
    }

    public record CameraDefinition(String id, String name,
                                    double focalLengthMm, double sensorWidthMm,
                                    double aperture, double nearClip, double farClip,
                                    String aspectRatioOverride) {}

    public record Quaternion(double x, double y, double z, double w) {
        public boolean isNormalized() {
            double norm = Math.sqrt(x*x + y*y + z*z + w*w);
            return Math.abs(norm - 1.0) < 1e-6;
        }
    }

    public record Vector3(double x, double y, double z) {}

    public record TimedTransform(int timeMs, Vector3 position, Quaternion rotation, Vector3 scale) {
        public TimedTransform {
            if (scale == null) scale = new Vector3(1, 1, 1);
        }
    }

    public record ActionClip(String clipKey, int inMs, int outMs, double weight) {}

    public record SceneObject(String id, String name, String type, String subType,
                               List<TimedTransform> keyframes, List<ActionClip> actions,
                               Vector3 position, Quaternion rotation, Vector3 scale,
                               String assetId, String groupId, boolean hidden) {}

    public record TimelineTrack(String trackId, String trackType, List<TimedKeyframe> keyframes) {}

    public record TimedKeyframe(int timeMs, String property, Object value, String interpolation) {}

    public record Timeline(List<TimelineTrack> tracks) {}
}
